trieste/space.py

```diff
@@ -920,17 +917,242 @@ def __init__(self, spaces: Sequence[SearchSpace], tags: Optional[Sequence[str]]
+            from each search subspace with shape '[num_samples, D]' , where D is the search space
+            dimension.
+        """
+        tf.debugging.assert_non_negative(num_samples)
+        if seed is not None:  # ensure reproducibility
+            tf.random.set_seed(seed)
+        return [self._spaces[tag].sample(num_samples, seed=seed) for tag in self._tags]
+
+    def __eq__(self, other: object) -> bool:
+        """
+        :param other: A search space.
+        :return: Whether the search space is identical to this one.
+        """
+        if not isinstance(other, type(self)):
+            return NotImplemented
+        return self._tags == other._tags and self._spaces == other._spaces
+
+
+class TaggedMultiSearchSpace(CollectionSearchSpace):
+    r"""
+    A :class:`SearchSpace` consisting of a collection of multiple :class:`SearchSpace` objects,
+    each with a unique tag. This class provides functionality for accessing each individual space.
+
+    Note that the individual spaces are not combined in any way, however all subspaces should have
+    the same dimension.
+    """
+
+    def __init__(self, spaces: Sequence[SearchSpace], tags: Optional[Sequence[str]] = None):
+        r"""
+        Build a :class:`TaggedProductSearchSpace` from a list ``spaces`` of other spaces. If
+        ``tags`` are provided then they form the identifiers of the subspaces, otherwise the
+        subspaces are labelled numerically.
+
+        :param spaces: A sequence of :class:`SearchSpace` objects representing the space's subspaces
+        :param tags: An optional list of tags giving the unique identifiers of
+            the space's subspaces.
+        :raise ValueError (or tf.errors.InvalidArgumentError): If ``spaces`` has a different
+            length to ``tags`` when ``tags`` is provided or if ``tags`` contains duplicates.
+        :raise ValueError (or tf.errors.InvalidArgumentError): If ``spaces`` has a different
+            dimension to each other.
+        """
+
+        # At least one subspace is required.
+        tf.debugging.assert_greater(
+            len(spaces),
+            0,
+            message=f"""
+                At least one subspace is required but received {len(spaces)}.
+                """,
+        )
+
+        tf.debugging.assert_equal(
+            len(set([int(space.dimension) for space in spaces])),
+            1,
+            message=f"""
+                All subspaces must have the same dimension but received
+                {[int(space.dimension) for space in spaces]}.
+                """,
+        )
+
+        super().__init__(spaces, tags)
+
+    def __repr__(self) -> str:
+        """"""
+        return f"""TaggedMultiSearchSpace(spaces =
+                {[self.get_subspace(tag) for tag in self.subspace_tags]},
+                tags = {self.subspace_tags})
+                """
+
+    @property
+    def lower(self) -> TensorType:
```

Comment:
Not sure how easily/accurately we can shape annotate this (though it might be possible, e.g. see the shapes of custom types section at https://gpflow.github.io/check_shapes/main/index.html) but we should at least docstring the behaviour, as it is counterintuitive.